### PR TITLE
Add TapSense Header Bidding Adapter and tests

### PR DESCRIPTION
**Pull request by erikchau**
## Type of change

- [x] New bidder adapter

## Description of change
Addition of the TapSense Header Bidding Adapter

- test parameters for validating bids
```
{
  bidder: 'tapsense',
  params: {
    ufid: "ZFiAg3MMs6E4KVyIDVbi8165506790392",
    refer: "localhost",
    jsonp: 1,
    ad_unit_id: "5877f75b0118065391b18f8c",
    device_id: "7bd153d3-8e9f-4bae-a4b3-751614d788b5",
    lat: "47.5643",
    long: "-122.151",
    scriptURL: "https://ads04.tapsense.com/ads/mopubad",
    user: "771fc655-6e4e-4bae-a4b6-751614d788b5",
    price_floor: 0.04,
    test: 1
  }
}
```
- contact email of the adapter’s maintainer
    amit@tapsense.com
- [x] official adapter submission

### CODE REVIEW

**Comment by erikchau** [comment]
Hi @mkendall07, we currently have a publisher waiting to integrate with us, anyway we can help to speed up the merge process?

tagging @amitmanjhi

File: src/adapters/tapsense.js [hunk 1]
```diff
@@ -0,0 +1,89 @@
+//v0.0.1
+var bidfactory = require('../bidfactory.js');
```
Comment (mkendall07) [inline src/adapters/tapsense.js]: please use `const`/`let` instead of `var` as appropriate.

File: src/adapters/tapsense.js [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+        if (creativeSizes.indexOf(bid.sizes[k].join("x")) > -1) {
+          isValidSize = true;
+          break;
+        }
+      }
+      if (isValidSize) {
+        if (!bid.params.scriptURL) {
+          continue;
+        }
+        var queryString = "?price=true&callback=tapsense.callback_with_price_" + bid.bidId + "&version=" + version + "&";
```
Comment (mkendall07) [inline src/adapters/tapsense.js]: Prefer string templates here instead of concatenation.

File: src/adapters/tapsense.js [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+        window.tapsense["callback_with_price_" + bid.bidId] = generateCallback(bid.bidId);
```
Comment (mkendall07) [inline src/adapters/tapsense.js]: use the the prebid global object `$$PREBID_GLOBAL$$` instead of window.

File: src/adapters/tapsense.js [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+        var keys = Object.keys(bid.params);
+        for (var j = 0; j < keys.length; j++) {
+          if (validParams.indexOf(keys[j]) < 0) continue;
+          queryString += encodeURIComponent(keys[j]) + "=" + encodeURIComponent(bid.params[keys[j]]) + "&";
+        }
+        var scriptURL = bid.params.scriptURL;
```
Comment (mkendall07) [inline src/adapters/tapsense.js]: can you make `scriptUrl` static or is it required to be passed in?

File: src/adapters/tapsense.js [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+        _requestBids(scriptURL + queryString);
+      }
+    }
+  }
+
+  function generateCallback(bidId){
+    return function(response, price) {
```
Comment (mkendall07) [inline src/adapters/tapsense.js]: Any reason this is an anonymous function?

File: src/adapters/tapsense.js [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+  function _callBids(params) {
+    bids = params.bids || [];
+    for (var i = 0; i < bids.length; i++) {
+      var bid = bids[i];
+      var isValidSize = false;
+      if (!bid.sizes) {
+        return;
+      }
+      for (var k = 0; k < bid.sizes.length; k++) {
+        if (creativeSizes.indexOf(bid.sizes[k].join("x")) > -1) {
```
Comment (mkendall07) [inline src/adapters/tapsense.js]: This fails if the `bid.sizes` is a single array and not a nested array (ie `[300,250];` instead of `[[300,250]];`)

**Comment by mkendall07** [comment]
@erikchau 
Can you also mention which params are required and which are optional? Thanks

**Comment by erikchau** [comment]
<table class="table table-bordered table-striped">
  <thead>
    <tr>
      <th style="text-align: left">Name</th>
      <th style="text-align: left">Scope</th>
      <th style="text-align: left">Description</th>
      <th style="text-align: left">Example</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td style="text-align: left"><code class="highlighter-rouge">ad_unit_id</code></td>
      <td style="text-align: left">required</td>
      <td style="text-align: left">Placement ID from TapSense</td>
      <td style="text-align: left">“ 5877f75b0118065391b18f8c”</td>
    </tr>
    <tr>
      <td style="text-align: left"><code class="highlighter-rouge">user</code></td>
      <td style="text-align: left">required</td>
      <td style="text-align: left">User's IDFA</td>
      <td style="text-align: left">"771fc655-6e4e-4bae-a4b6-751614d788b5"</td>
    </tr>
    <tr>
      <td style="text-align: left"><code class="highlighter-rouge">refer</code></td>
      <td style="text-align: left">optional</td>
      <td style="text-align: left">Site that ad request originated from</code>
</td>
      <td style="text-align: left">“https://tapsense.com”</td>
    </tr>
    <tr>
      <td style="text-align: left"><code class="highlighter-rouge">lat</code></td>
      <td style="text-align: left">optional</td>
      <td style="text-align: left">Users latitude
</td>
      <td style="text-align: left">“ 47.5643”</td>
    </tr>
<tr>
      <td style="text-align: left"><code class="highlighter-rouge">long</code></td>
      <td style="text-align: left">optional</td>
      <td style="text-align: left">Users longitude
</td>
      <td style="text-align: left">“-122.151”</td>
    </tr>
<tr>
      <td style="text-align: left"><code class="highlighter-rouge">price_floor</code></td>
      <td style="text-align: left">optional</td>
      <td style="text-align: left"> Minimum Bid in dollars
</td>
      <td style="text-align: left">“ 0.04”</td>
    </tr>
  </tbody>
</table>

**Comment by erikchau** [comment]
@mkendall07 ready for re-review

File: src/adapters/tapsense.js [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+
+const bidfactory = require('../bidfactory.js');
+const bidmanager = require('../bidmanager.js');
+const adloader = require('../adloader');
+const utils = require('../utils.js');
+
+const TapSenseAdapter = function TapSenseAdapter() {
+  const version = "0.0.1";
+  const creativeSizes = [
+    "320x50"
```
Comment (mkendall07) [inline src/adapters/tapsense.js]: This is the only valid bid size?

**Comment by mkendall07** [comment]
Thanks for the updates. With the changes, I'm getting 
```pbjs.tapsense.callback_with_price_26c264cf22f421({"id": "58becf1bda78506f6409a5fb","status": {"value": "error"},"count_ad_units": 0}, 1.2)```
Response with specified bid params.

**Comment by erikchau** [comment]
Sorry, should have mentioned that the ad unit id I supplied was for android devices. You will need to set your UA as any android device. Our servers check that the user agent matches the ad unit's platform.
screen shot 2017-03-07 at 10 48 21 am

File: src/adapters/tapsense.js (hunk 7 above)
Comment (erikchau) [inline src/adapters/tapsense.js]: For header bidding we only currently only support banner requests

Comment (mkendall07) [inline src/adapters/tapsense.js]: ok, probably want to make a note about this in your bidder documentation.

Comment (erikchau) [inline src/adapters/tapsense.js]: Just to be sure, to create our bidder documentation I will need to make a PR for this page with our information correct?? http://prebid.org/dev-docs/bidders.html

**Comment by erikchau** [comment]
Created bidder documentation PR here: https://github.com/prebid/prebid.github.io/pull/186
Is there anything else left to be done on my end @mkendall07 ?
thanks

**Comment by mkendall07** [comment]
LGTM. Thanks

**Comment by mkendall07** [comment]
@erikchau 
This is merged into master. Please submit a PR to the [docs repo](https://github.com/prebid/prebid.github.io) to add a file for your adapter to the [bidders directory](https://github.com/prebid/prebid.github.io/tree/master/dev-docs/bidders) so your adapter's params will appear on the [bidders page](http://prebid.org/dev-docs/bidders.html). Thank you for contributing

**Comment by mkendall07** [comment]
oh, great. You were fast :) thanks

**Pull request merged**